Cross-platform support of Vorbis/MP3 music files.

dexsda opened the pull request:
The changes require the flags `USE_CODEC_VORBIS` and `USE_CODEC_MP3` to be set during compilation.

FMOD has been removed in this PR. The files will work from inside a pak file. mp3 files will take precedence over ogg files.

Libraries `libmad` and `libvorbis` are required; for Windows, they can be found here: https://github.com/shalrathy/quakespasm-shalrathy/tree/master/Windows/codecs/x86. The linux fork by @matthewearl will contain README.md changes that reflect what is necessary there.

j0zzz commented:
I was able to compile it on Windows and it works nicely.
I had the following 2 compiler errors (the visual studio project is set to 'treat warnings as errors'):

- in **BGM_PlayCDtrack** the variable `path_id` has no initial value
Tried to set it `0` as initial value, but then music playback doesn't start, due to the condition `if (path_id > prev_id)` would never be true. So it seems like it must have a > 0 initial value, which might be true by default on Linux, but we should give a proper initial value to it, instead of some memory garbage that accidentally satisfies the condition.

- in **S_CodecUtilOpen** the method `COM_FOpenFile` is called with `filename` param which has `const char *` type, however `COM_FOpenFile` expects `char *`. I simply used type cast to `(char *)` to shut up the compiler.

I also found a bug:

- When you toggle the `bgm_extmusic` cvar, the change doesn't take place if you then start such level that plays the same audio track. This happens due to the start of the `BGM_PlayCDtrack` method. I understand the optimization to keep running the current track, but then there should be another check that if the cvar was toggled, playback should be stopped at any case. I am willing to fix this using an `onChange` cvar function, where at the same time I would also stop playback immediately, as it was with the previous `fmod_music` cvar.

I also need to include 4 additional .dll's from now in the JoeQuake win32 release package, but it's the same with QuakeSpasm, so I guess we're just gonna do it like that. If it's better for Linux users, I don't really mind.

dexsda commented:
Regarding suggestion 1, you're right: this is a feature in ironwail. Unfortunately, it requires substantial changes to code to work. My vote is to remove the feature with path_id entirely and just play in the order of music handlers. FMOD previously had no such feature either so we're not losing anything by removing that.

Regarding the second, that type cast is fine. But there's a change I made to use const in one function definition which, if we choose to use a type cast to shut up the compiler, probably should be reverted.

I can fix both of these sometime tomorrow. You mentioned you'll work on the onChange function to fix the bug, so I'll leave that to you.

j0zzz commented:
> I can fix both of these sometime tomorrow. You mentioned you'll work on the onChange function to fix the bug, so I'll leave that to you.

Sounds good to me. I noticed another issue though: the bgmvolume is very quiet now compared to previously, and also it is a lot more loud in QuakeSpasm. You can barely hear the background music even with setting bgmvolume to 1.0.

dexsda commented:
The problem with volume is that the current volume for sounds is brought all the way up, and they like to clip. Quakespasm solves this problem by implementing a low-pass filter on the samples which makes sounds about half as loud and removes clipping. Then, music can be safely mixed at the required volume.

There are three ways to fix the issue:
 - juice up the music by bitshifting samples more. This will start to clip and sound awful, though; in my experimentation the value I've taken was pretty good and the same as the samples for sounds, but we could try to bring it up by one more bitshift. It might still be fine but should double the volume exactly.
 - bring down the sounds a bit by bitshifting samples for sounds less. This would bring it more in line with music, but now the settings on the sounds might feel a little less loud. Both these solutions are hacky.
 - port the low-pass filter from quakespasm. This will make the sounds sound a lot better overall and fix the music volume issue. But it's a bit of work.

To be honest I wanted to do solution 3 at some point *anyway*, and could produce one. But I won't have the time to sit down on it for another week. If we're fine delaying the merge a little bit on this I can pledge to do work on it after my university teaching obligations are finished for the semester.

j0zzz commented:
> BTW right now I fix the issue by just having sounds at 0.5 and bgm at 1.0 volume in my linux build. This is equivalent to exactly a bitshift difference.

Well that's good for you, but all the JoeQuake users will experience this as a bug if they stored music settings aren't work the way it did previously. I need to have this working exactly the same as before.

Anyways, I stashed my local changes and will apply them when you fixed the PR.

dexsda commented:
> > BTW right now I fix the issue by just having sounds at 0.5 and bgm at 1.0 volume in my linux build. This is equivalent to exactly a bitshift difference.
> 
> Well that's good for you, but all the JoeQuake users will experience this as a bug if their stored music settings aren't work the way it did previously. I need to have this working exactly the same as before.
> 
> Anyways, I stashed my local changes and will apply them when you fixed the PR.

Understood, just giving a rationale for the bitshifting difference as a possible solution :).

dexsda commented:
Added the fixes and ported the entirety of snd_mix from ironwail (except for the underwater effect...). I've fixed some of the casting issues.

The sounds are now linearly interpolated so there should be less clipping when there's lots of them. 11khz and 44 khz should work much better together now, though there is no way to gain any quality on the 11khz samples of course.

There should be more headroom for music now, and sound mixing should be handled better independently. The old asm is no longer used for the mixing; by now the compilers probably generate equally good assembly from the C code anyway.

Once you apply your stashed changes we should be good to go.

j0zzz commented:
After merging and applying my stash, the game immediately crashes after startup (GL Debug/Release versions):
![image](https://github.com/user-attachments/assets/6009cee1-9085-4993-baae-3f2976b34b3b)
The issue is coming from the new code checked in with commit 1fa3088.

The SDL version doesn't crash - I assume it's not calling this method. But the music volume is still very low compared to the FMOD version.

dexsda commented:
I'll test some music volume boosting values, with the interpolation we have headroom to avoid clipping.

The error happens in the GL version because the `snd_out` buffer is allocated to be `shm->buffer` in the code. The SDL code runs a run-of-the-mill `calloc`, and works on the Linux version since it's SDL only. The GL version on windows, I suspect, runs the `GlobalAlloc` function instead; the allocated size must be too low. I'll dig into this a bit.

dexsda commented:
Doubled the volume on the music, it appears not to clip so this should have the music in better parity with what it was in FMOD.

Still haven't looked very closely at the reason for the error, since it's hard for me to test non-SDL builds. Would it be possible for you to provide what the value of `i` is at the time of the crash?

dexsda commented:
In the SDL part, the `shm->buffer` is assigned with a `calloc` in the new code. This is why it works in Linux (where SDL is the only build) and in the SLD Windows build. In the windows related part, it is assigned differently. That is alien to me. But I am surprised that it isn't assigned at all, I expected too small a buffer. I'll also take a look at the differences in the new code that might cause this.

dexsda commented:
In the `snd_mix.c` file, the old version of the code has this:

```
#if defined(_WIN32) && !defined(SDL2)
	if (pDSBuf)
	{
		reps = 0;

		while ((hresult = pDSBuf->lpVtbl->Lock(pDSBuf, 0, gSndBufSize, &pbuf, &dwSize, &pbuf2, &dwSize2, 0)) != DS_OK)
		{
			if (hresult != DSERR_BUFFERLOST)
			{
				Con_Printf ("S_TransferStereo16: DS::Lock Sound Buffer Failed\n");
				S_Shutdown ();
				S_Startup ();
				return;
			}

			if (++reps > 10000)
			{
				Con_Printf ("S_TransferStereo16: DS: couldn't restore buffer\n");
				S_Shutdown ();
				S_Startup ();
				return;
			}
		}
	}
	else
#endif
	{
		pbuf = (DWORD *)shm->buffer;
	}
```
and `pbuf` is then assigned to snd_out. So it seems there is a bit of code specifically to handle issues in the windows build.

It might be necessary to restore something like this for the Windows build to work correctly.

j0zzz commented:
Yeah, please do and add it to the PR, I'll then test it out.

dexsda commented:
Added the code back, and fixed a bug in the SDL sound mixing from JoeQuake while I was at it.

dexsda commented:
Restored the code, and fixed an SDL bug while I was at it.
